crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1655,37 +1650,68 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = setcomp;
 
         self.infer_expression(elt);
-        for comprehension in generators {
-            self.infer_comprehension(comprehension);
-        }
+        self.infer_comprehensions_scope(generators);
     }
 
-    fn infer_comprehension(&mut self, comprehension: &ast::Comprehension) {
-        self.infer_definition(comprehension);
-        for expr in &comprehension.ifs {
-            self.infer_expression(expr);
+    fn infer_comprehensions_scope(&mut self, comprehensions: &[ast::Comprehension]) {
+        let mut generators_iter = comprehensions.iter();
+        let Some(first_generator) = generators_iter.next() else {
+            unreachable!("Comprehension must contain at least one generator");
+        };
+        self.infer_comprehension_scope(first_generator, true);
+        for generator in generators_iter {
+            self.infer_comprehension_scope(generator, false);
```

Comment:
In the parallel code above in `infer_first_comprehension_iter` you changed all the naming from `generator*` to `comprehension*`, which is reasonable since `generators` attribute is a list of `Comprehension` nodes. Let's at least be consistent and use the same naming both here and in `infer_first_comprehension_iter`.